pdata/pmetric/json_test.go

```diff
@@ -164,6 +167,290 @@ func TestMetricsNil(t *testing.T) {
+	datapoint.SetStartTimestamp(internal.NewTimestampFromTime(time.Now()))
+	datapoint.SetFlags(internal.MetricDataPointFlagsNone)
+	datapoint.SetCount(100)
+	datapoint.SetSum(100)
+	quantile := datapoint.QuantileValues().AppendEmpty()
+	quantile.SetQuantile(0.5)
+	quantile.SetValue(1.2)
+	datapoint.Attributes().UpsertString("string", "value")
+	datapoint.Attributes().UpsertBool("bool", true)
+	datapoint.Attributes().UpsertInt("int", 1)
+	datapoint.Attributes().UpsertDouble("double", 1.1)
+	datapoint.Attributes().UpsertMBytes("bytes", []byte("foo"))
+	datapoint.SetTimestamp(internal.NewTimestampFromTime(time.Now()))
+	return metric
+}
+
+func TestMetricsSumDataJSONFull(t *testing.T) {
+	m := metricsSumOTLPFull()
+	encoder := NewJSONMarshaler()
+	jsonBuf, err := encoder.MarshalMetrics(m)
+	assert.NoError(t, err)
+	decoder := NewJSONUnmarshaler()
+	got, err := decoder.UnmarshalMetrics(jsonBuf)
+	assert.NoError(t, err)
+	assert.EqualValues(t, m, got)
+}
+
+func TestMetricsGaugeDataJSONFull(t *testing.T) {
+	m := metricsGaugeOTLPFull()
+	encoder := NewJSONMarshaler()
+	jsonBuf, err := encoder.MarshalMetrics(m)
+	assert.NoError(t, err)
+	decoder := NewJSONUnmarshaler()
+	got, err := decoder.UnmarshalMetrics(jsonBuf)
+	assert.NoError(t, err)
+	assert.EqualValues(t, m, got)
+}
+
+func TestMetricsHistogramDataJSONFull(t *testing.T) {
+	m := metricsHistogramOTLPFull()
+	encoder := NewJSONMarshaler()
+	jsonBuf, err := encoder.MarshalMetrics(m)
+	assert.NoError(t, err)
+	decoder := NewJSONUnmarshaler()
+	got, err := decoder.UnmarshalMetrics(jsonBuf)
+	assert.NoError(t, err)
+	assert.EqualValues(t, m, got)
+}
+
+func TestMetricsExponentialHistogramDataJSONFull(t *testing.T) {
+	m := metricsExponentialHistogramOTLPFull()
+	encoder := NewJSONMarshaler()
+	jsonBuf, err := encoder.MarshalMetrics(m)
+	assert.NoError(t, err)
+	decoder := NewJSONUnmarshaler()
+	got, err := decoder.UnmarshalMetrics(jsonBuf)
+	assert.NoError(t, err)
+	assert.EqualValues(t, m, got)
+}
+
+func TestMetricsSummaryDataJSONFull(t *testing.T) {
+	m := metricsSummaryOTLPFull()
+	encoder := NewJSONMarshaler()
+	jsonBuf, err := encoder.MarshalMetrics(m)
+	assert.NoError(t, err)
+	decoder := NewJSONUnmarshaler()
+	got, err := decoder.UnmarshalMetrics(jsonBuf)
+	assert.NoError(t, err)
+	assert.EqualValues(t, m, got)
+}
+
+func TestReadKvlistValueUnknownField(t *testing.T) {
+	jsonStr := `{"extra":""}`
+	iter := jsoniter.ConfigFastest.BorrowIterator([]byte(jsonStr))
+	defer jsoniter.ConfigFastest.ReturnIterator(iter)
+	readKvlistValue(iter)
+	if assert.Error(t, iter.Error) {
+		assert.Contains(t, iter.Error.Error(), "unknown field")
+	}
+}
+
+func TestReadArrayUnknownField(t *testing.T) {
+	jsonStr := `{"extra":""}`
+	iter := jsoniter.ConfigFastest.BorrowIterator([]byte(jsonStr))
+	defer jsoniter.ConfigFastest.ReturnIterator(iter)
+	readArray(iter)
+	if assert.Error(t, iter.Error) {
+		assert.Contains(t, iter.Error.Error(), "unknown field")
+	}
+}
```

Comment:
It will be better to add benchmark compare to jsonpb.

Reply:
I will do it